### ISSUE TITLE
Inconsistent ruby script behavior

### ISSUE THREAD

**Issue by GladB** [databrary/datavyu#215]
**Describe the bug**
There is a ruby script that I often run, and it's behavior is random: without changing anything, it will sometimes run without issue and sometimes throw an error.

Running without issue, I get:

```  
*************************
Running Script: run_all_postannotation.rb on project: (untitled)
*************************


Script has finished running. 
```

But I sometimes get the following:

```
Running Script: run_all_postannotation.rb on project: (untitled)
*************************
NoMethodError: undefined method `getAllVariables' for nil:NilClass
  get_column_list at file:/Applications/datavyu.app/Contents/Java/datavyu-1.3.7-jfx.jar!/Datavyu_API.rb:2358
           (root) at <script>:55


***** SCRIPT ERROR *****

NoMethodError: undefined method `getAllVariables' for nil:NilClass
  get_column_list at file:/Applications/datavyu.app/Contents/Java/datavyu-1.3.7-jfx.jar!/Datavyu_API.rb:2358
           (root) at <script>:55

See line 55 of /Volumes/pn-opus/Seedlings/Scripts_and_Apps/Github/seedlings/datavyu/run_all_postannotation.rb:
	columns = get_column_list()
************************* 
```

I sometimes also have the following:

```
Errno::EBADF: Bad file descriptor - Bad file descriptor close at org/jruby/RubyIO.java:2101 readlines at org/jruby/RubyIO.java:4062 (root) at :13
See line 13 of /Volumes/pn-opus/Seedlings/Scripts_and_Apps/Github/seedlings/datavyu/run_all_postannotation.rb: $valid_id_newline = File.readlines("/Volumes/pn-opus/Seedlings/usedID.txt")
```

or 

```
NoMethodError: undefined method `include?' for nil:NilClass checkCodes at :19 (root) at :84 each at org/jruby/RubyArray.java:1613 (root) at :77 each at org/jruby/RubyArray.java:1613 (root) at :72
* SCRIPT ERROR * org.jruby.embed.EvalFailedException: (NoMethodError) undefined method `include?' for nil:NilClass
```

```
NoMethodError: private method `getVariable' called for nil:NilClass get_column at file:/Applications/datavyu 2.app/Contents/Java/datavyu-1.3.7-jfx.jar!/Datavyu_API.rb:722 (root) at :256 each at org/jruby/RubyArray.java:1613 (root) at :254
See line 254 of /Volumes/pn-opus/Seedlings/Scripts_and_Apps/Github/seedlings/datavyu/run_all_postannotation.rb: for column in columns See line 256 of /Volumes/pn-opus/Seedlings/Scripts_and_Apps/Github/seedlings/datavyu/run_all_postannotation.rb: col = get_column(column)
```

Note that I got all those errors simply by running the same script again and again without changing anything in between.

**To Reproduce**

I am myself not able to reproduce this behavior whenever I want. The steps that I follow to sometimes get this error are the following:
1. Modify lines 11 and 13 to point to the linked txt file
2. Open the linked opf file
3. Click on Script, Run script, select the ruby script that I linked here
4. See (one of the) error(s) (maybe?)

**Expected behavior**
This script should print the errors that are in the current opf file.

**Desktop (please complete the following information):**
 - OS: tested on macOS 10.13.6 and Windows 7
 - Version: tested with Datavyu 1.3.7 and 1.4.2, Ruby 2.3.7

[files_issue.zip](https://github.com/databrary/datavyu/files/3044391/files_issue.zip)

**Comment by Xubes** [comment]
Thanks for reporting this and providing the examples. 

I'm unable to reproduce the bad file descriptor error after several thousand iterations. I've noticed that error show up when the file is open in any other application, so I'd make sure that no other application is using the file. Other than that, I'm not sure what to do if it continues to show up.

We're aware of the getAllVariables() error popping up randomly and have tried to figure out why exactly it occurs. We think it's a JRuby issue. I'm able to reproduce it in 1.3.7 with the example you sent, but not in 1.4.2. In general, I would avoid using global variables (e.g., $my_var). I've noticed that they sometimes just become nil for no apparent reason. This was happening with the $db variable in the get_column_list function, which was causing the getAllVariables() error.

See the attached modified script which replaces the global vars. I've been able to successfully run this without errors in 1.4.2.

[run_all_postannotation2.rb.zip](https://github.com/databrary/datavyu/files/3055373/run_all_postannotation2.rb.zip)
